app/services/orders/create_product_orders.rb

```diff
@@ -0,0 +1,17 @@
+class Orders::CreateProductOrders
+  attr_reader :products_hash, :order
+
+  def initialize(products_hash, order)
+    @products_hash = products_hash
+    @order = order
+  end
+
+  def call
+    products_hash.each do |product_id, amount|
+      product = Product.find(product_id)
+      amount = [amount, product.balance].min
+
+      order.product_orders.create(product_id:, amount:)
+    end
```

Comment:
again, look, if you'll have 100000000 records, you make 100000000 calls to the DB, when you can do only one